librz/core/cgraph.c

```diff
@@ -0,0 +1,726 @@
+		core_graph_fn_call(core, fcn, graph);
+	}
+	return graph;
+}
+
+typedef char *(*GraphBodyFn)(RzCore *core, ut64 addr, RzAnalysisBlock *bb);
+
+static inline char *block_disasm(RzCore *core, ut64 addr, RzAnalysisBlock *bb) {
+	return rz_core_cmd_strf(core, "pdb @ 0x%" PFMT64x, addr);
+}
+
+static inline char *block_diff(RzCore *core, ut64 addr, RzAnalysisBlock *bb) {
+	if (!bb) {
+		bb = rz_analysis_get_block_at(core->analysis, addr);
+		if (!bb) {
+			return NULL;
+		}
+	}
+
+	char *str_a = block_disasm(core, bb->addr, NULL);
+	if (!bb->diff) {
+		return str_a;
+	}
+	char *str_b = block_disasm(core, bb->diff->addr, NULL);
+	RzDiff *diff = rz_diff_lines_new(str_a, str_b, NULL);
+	char dff_from[32], dff_to[32];
+	snprintf(dff_from, sizeof(dff_from), "0x%08" PFMT64x, bb->addr);
+	snprintf(dff_to, sizeof(dff_to), "0x%08" PFMT64x, bb->diff->addr);
+	char *body = rz_diff_unified_text(diff, dff_from, dff_to, false, false);
+	rz_diff_free(diff);
+
+	free(str_a);
+	free(str_b);
+	return body;
+}
+
+static inline RzGraphNode *graph_add_cached(RzCore *core, HtUP *cache, RzAnalysisBlock *bb, ut64 offset, RzGraph *graph, GraphBodyFn body_fn) {
+	RzGraphNode *node = (RzGraphNode *)ht_up_find(cache, offset, NULL);
+	if (node) {
+		return node;
+	}
+
+	char *title = rz_address_str(offset);
+	char *body = body_fn(core, offset, bb);
+	node = rz_graph_add_node_info(graph, title, body, offset);
+	ht_up_insert(cache, offset, node);
+	free(title);
+	free(body);
+	return node;
+}
+
+static void core_graph_fn_bbs(RzCore *core, RzAnalysisFunction *fcn, RzGraph *graph, HtUP *cache, GraphBodyFn body_fn) {
+	if (!(fcn && fcn->bbs)) {
+		return;
+	}
+
+	RzListIter *iter;
+	RzAnalysisBlock *bbi;
+	rz_list_foreach (fcn->bbs, iter, bbi) {
+		if (bbi->addr == UT64_MAX) {
+			continue;
+		}
+
+		RzGraphNode *bb_node = graph_add_cached(core, cache, bbi, bbi->addr, graph, body_fn);
+		if (!bb_node) {
+			continue;
+		}
+
+		if (bbi->jump != UT64_MAX) {
+			RzGraphNode *node = graph_add_cached(core, cache, NULL, bbi->jump, graph, body_fn);
+			if (node) {
+				rz_graph_add_edge(graph, bb_node, node);
+			}
+		}
+
+		if (bbi->fail != UT64_MAX) {
+			RzGraphNode *node = graph_add_cached(core, cache, NULL, bbi->fail, graph, body_fn);
+			if (node) {
+				rz_graph_add_edge(graph, bb_node, node);
+			}
+		}
+
+		if (bbi->switch_op) {
+			RzAnalysisCaseOp *case_op;
+			RzListIter *iter_case;
+			rz_list_foreach (bbi->switch_op->cases, iter_case, case_op) {
+				RzGraphNode *case_node = graph_add_cached(core, cache, NULL, case_op->addr, graph, body_fn);
+				if (case_node) {
+					rz_graph_add_edge(graph, bb_node, case_node);
+				}
+
+				RzGraphNode *case_jump_node = graph_add_cached(core, cache, NULL, case_op->jump, graph, body_fn);
+				if (case_jump_node) {
+					rz_graph_add_edge(graph, case_node, case_jump_node);
+				}
+			}
+		}
+	}
+}
+
+static inline RzGraph /*<RzGraphNodeInfo *>*/ *rz_core_graph_function_bbs(RzCore *core, ut64 addr, GraphBodyFn body_fn) {
```

Comment:
No need for `inline` here. Moreover, while it's static, it still deserves a short Doxygen comment about the purpose - it's unclear from the function name itself.